handout/zones/README.md

```diff
@@ -0,0 +1,132 @@
+    this.ngZone.onError.subscribe(this.onZoneError);
+  }
+
+  onZoneStable() {
+    console.log('We are stable');
+  }
+
+  onZoneUnstable() {
+    console.log('We are unstable');
+  }
+
+  onZoneError(error) {
+    console.error('Error', error instanceof Error ? error.message : error.toString());
+  }
+}
+```
+
+Subscribing to these can help you determine if your code is unexpectedly triggering change detection as a result of operations that do not affect application state.
+
+## Change Detection
+
+Since all asynchronous code executed from within Angular's zone can trigger change detection you may prefer to execute some code outside of Angular's zone when change detection is not required.
+
+To run code outside of Angular's context, **NgZone** provides a method aptly named **runOutsideAngular**. Using this method, Angular's zone will not interact with your code and will not receive events when the global zone becomes stable.
+
+In this [example](http://plnkr.co/edit/jhnTaC?p=preview) you will see in the console what happens with Angular's zone when code is run in and outside of it.
+
+You will notice that in both cases clicking the button causes the Angular zone to become unstable due to Zone.js patching and watching **HTMLElement.prototype.onclick**, however the **setInterval** executing outside of Angular's zone does not affect its stability and does not trigger change detection.
+
+## Debugging
+
+Generally, exceptions thrown during a chain of asynchronous events will only include the current method in their stack trace.
+
+Since Zone.js tracks all of our asynchronous calls it can provide us a longer, more detailed, stack trace of the events and calls that occurred leading up to our exception.
+
+To enable long stack traces in development, you should include the **long-stack-trace-zone** module in your code. It is a good idea not to include this in your production build but Angular will skip setting up longer stack traces when in production mode (`enableProdMode` from `@angular/core`).
+
+Angular will take care of forking and extending its own zone to display more meaningful stack traces.
+
+```js
+if (__PRODUCTION__) {
+  enableProdMode();
+} else {
+  require('zone.js/dist/long-stack-trace-zone');
+}
+```
+
+With the following code, we start by calling `startAsync` which triggers a chain of setTimeouts leading up to an uncaught error.
+
+```js
+function startAsync() {
+  setTimeout(stepOne, 100);
+}
+
+function stepOne() {
+  setTimeout(stepTwo, 100);
+}
+
+function stepTwo() {
+  throw new Error('Finished');
+}
+
+```
+
+### Simple Stack trace
+
+This is a typical stack trace that you would see in this scenario, without Zone.
+
+```
+Uncaught Error: Finished(…)
+  stepTwo @ debugging.html:28
+```
+
+### Detailed "Long" Stack trace
+
+In the stack trace below, you can determine the order of function calls and events that occurred within this asynchronous chain of events ('>' added for emphasis).
+
+```
+debugging.html:16 Error: Finished
```

Comment:
We should probably explain what the benefit of the long stack trace is. At first glance it looks like most of the stack trace is zone stuff. It took me a while to notice `stepOne` and `startAsync`.